### DISCUSSION TITLE
Clarification needed regarding d=i and d=I and keeping images in memory when they are hidden.

### DISCUSSION THREAD

**Discussion by vurentjie** [kovidgoyal/kitty#8129] [Q&A] [answered]
I need some clarification how to keep hidden images in memory while deleting others? Is this possible?

---

As a simple example. If I have two images. 

I show the first one. And then I hide it with `d=i` so that I do not have to retransmit it.

I add a second one. And then I hide it with `d=I` because I do want to free it. 

If the first image is hidden while I delete the other one with `d=I`, it is also removed and needs to be retransmitted. 
   
---

In the example below, at step 6 and step 7, I need some clarification. 

If I omit step 6 (and keep image 1 visible), then when deleting image 2 (in step 7) it does not affect image 1. 

```bash
img_1=$(printf "/home/vurentjie/Downloads/test1.png" | base64)
img_2=$(printf "/home/vurentjie/Downloads/test2.png" | base64)

# Step 1: transmit and store img with i=1
printf '\e_Gt=f,q=2,f=100,a=t,i=1;'$img_1'\e\\'

# Step 2: show image i=1 and p=2
printf '\e[?2026h'
printf '\e[s'
printf '\e[1;1H'
printf '\e_Gi=1,C=1,a=p,z=1,c=4,r=4,p=2,q=2;\e\\'
printf '\e[u'
printf '\e[?2026l'

# Step 3: remove image with i=1 (without free d=i)
printf '\e_Gq=2,i=1,d=i,a=d;\e\\'

# Step 4: show image i=1 and p=2 again (no need to retransmit)
# At this point this works as I expect it would because d=i
printf '\e[?2026h'
printf '\e[s'
printf '\e[1;1H'
printf '\e_Gi=1,p=2,a=p,z=1,r=4,C=1,q=2;\e\\'
printf '\e[u'
printf '\e[?2026l'

# Add the second image
# Step 5: transmit and store img with i=2
printf '\e_Gt=f,q=2,f=100,a=t,i=2;'$img_2'\e\\'

# Step 6: hide the image with i=1 (without free d=i)  ... note this is "Step 3" again, and I would expect "Step 8" to work again
printf '\e_Gq=2,i=1,d=i,a=d;\e\\'

# Step 7: now remove image with i=2 (but with free d=I)
# Somehow this also removes i=1 because it was hidden in "Step 6"
printf '\e_Gq=2,i=2,d=I,a=d;\e\\'

# Step 8: try show image i=1 and p=2 again
# This does not work, if both "Step 6 and Step 7" are also run 
printf '\e[?2026h'
printf '\e[s'
printf '\e[1;1H'
printf '\e_Gi=1,p=2,a=p,z=1,r=4,C=1,q=2;\e\\'
printf '\e[u'
printf '\e[?2026l'



```
Am I missing anything obvious here?

**Comment by kovidgoyal** [comment, marked as answer]
https://github.com/kovidgoyal/kitty/commit/8facaf4fa02822e6f9dc3597fea57fa951f8865b

**Comment by vurentjie** [comment]
Hi @kovidgoyal 

Thanks for fixing the other issue. I know I marked this thread as answered already, but I have another query.  

I am trying to understand how to free up storage manually. Is it correct to assume?
- Transmit and add image >> `GraphicsManager->used_storage` increases 
- Remove and free image (d=I) >>  `GraphicsManager->used_storage` decreases

---

Again with a simplified example:  I transmit a single image, and then I remove it with d=I. 

Example:
```bash
img_1=$(printf "/home/vurentjie/Downloads/test1.png" | base64 -w0)
# Transmit and store img with i=1
printf '\e_Gt=f,q=2,f=100,a=t,i=1;'$img_1'\e\\'
# Remove and free the img with i=1
printf '\e_Gq=2,i=1,d=I,a=d;\e\\'
```
---
I was adding a few logs in kitty/graphics.c to check the value of `GraphicsManager->used_storage` and I noticed it was not changing after the delete command.

I am not sure if it is meant to change, if I am misunderstanding anything. 

I noticed that in `filter_refs` this particular line would free the image resource if it was run. But it does not run with the above example. Is that correct?
https://github.com/kovidgoyal/kitty/blob/5994a0bb2d42c950e5cda014fbd54a3af49c55fc/kitty/graphics.c#L1907

My logs showed me the following:
```
called when adding the image: finish_command_response: storage size: 20587392, i: 1 
called when deleting: filter_refs, storage: 20587392 (!free_only_matched || matched):n, !vt_size(&img->refs_by_internal_id):y, (free_images || img->client_id == 0):y 
called when deleting: handle_delete_command: storage size: 20587392, i: 1
`

**Reply by kovidgoyal** [reply]
https://github.com/kovidgoyal/kitty/commit/24a195c7c7f73b7cee8691254341e70709e5e232
